Support wrapping jar in a .bat file on MS-Windows

ikappaki opened the pull request:
Hi,

could you please review patch to use the same technique to generate self contained output script, but this time as a batch file executable on MS-Windows, addresses #2.

I've noticed that the original bash preamble also included the preamble for MS-Windows which I separated out to its own branch.

The resulting file on MS-Windows will have a `.bat` extension.

I've tested this to work in isolation both on MS-Windows and GNU/Linux.

Thanks

ikappaki committed:
Support wrapping jar in a .bat file on MS-Windows

ericdallo reviewed (commented):
Oh that's awesome @ikappaki ! is there a way to have a test for that somehow?

ikappaki committed:
Add jar tests and enable GH CI action across macos/ubuntu/win

The jar tests cover script invocation, jvm options, argument passing
and exit value propagation.

It also adds a new :bin-path key pointing to the resulting executable.

ericdallo commented:
@ikappaki thanks for the tests and the CI configuration bonus ❤️ 

Not sure if you are aware, running `bb test` from my linux I get:

```bash
Running tests in #{"test"}                                                                                                                                                                    
WARNING: seqable? already refers to: #'clojure.core/seqable? in namespace: clojure.core.incubator, being replaced by: #'clojure.core.incubator/seqable?                                       
WARNING: seqable? already refers to: #'clojure.core/seqable? in namespace: clostache.parser, being replaced by: #'clojure.core.incubator/seqable?                                             
                                                                                                                                                                                              
Testing deps-bin.impl.bin-test                                                                                                                                                                
Creating standalone executable: some-bin                                                                                                                                                      
Re-aligning zip offsets...                                                                                                                                                                    
Creating standalone executable: /tmp/c6ee9de5-8f9c-4880-bc13-a46bcce769d52683342575197417312/testbin                                                                                          
                                                                                                                                                                                              
FAIL in (binary) (bin_test.clj:45)                                                                                                                                                            
jvm opts passing                                                                                                                                                                              
expected: (= 0 (:exit proc))                                                                                                                                                                  
  actual: (not (= 0 1))                                                                                                                                                                       

FAIL in (binary) (bin_test.clj:46)
jvm opts passing
expected: (= [":prop xyz :v 5" ":prop zyx :v 0"] (str/split-lines (str out)))
  actual: (not (= [":prop xyz :v 5" ":prop zyx :v 0"] [""]))
Creating standalone executable: /tmp/42c94d2c-9f26-4de9-b340-ff24746a8abb4839571082548560559/testbin

FAIL in (binary) (bin_test.clj:61)
return error value in case of error
expected: (= [":prop testjar.return :v 7"] (str/split-lines (str out)))
  actual: (not (= [":prop testjar.return :v 7"] [""]))

FAIL in (binary) (bin_test.clj:63)
return error value in case of error
expected: (= 7 (:exit proc))
  actual: (not (= 7 1))

Ran 2 tests containing 10 assertions.
4 failures, 0 errors.
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message: 
```

ikappaki committed:
Add Emacs .dir-locals project settings

ikappaki commented:
HI @ericdallo

> Oh that's awesome @ikappaki ! is there a way to have a test for that somehow?

Yes, sorry I've missed the fact there were tests already defined for this lib, I thought it was test-less :)

> @ikappaki thanks for the tests and the CI configuration bonus ❤️
> Not sure if you are aware, running `bb test` from my linux I get:
> 
> ```shell
> ...                                                                                                                                                                                              
> FAIL in (binary) (bin_test.clj:45)                                                                                                                                                            
> jvm opts passing                                                                                                                                                                              
> expected: (= 0 (:exit proc))                                                                                                                                                                  
>   actual: (not (= 0 1))                                                                                                                                                                       

sorry that was a premature check in that should have targeted the wip branch.

I've squashed that out with a new check in that
1. Provides a new jar test covering jvm options, jar arguments and exit value.
2. Created a new bb script in `test\test-jar` and associated java src file to create a jar to use in the above. The resulting jar is checked in so we don't have to recreate it each time. The maintainers can do this manually if the needs arise.
3. The resulting script has a `.bat` extension on windows. I've updated the documentation to indicate that as well the fn in concern to return an extra `bin-path` key to point to he full name of the resulting executable. This is useful for testing.
4. an error was discovered in the bat template that did not return the exit value to the caller. I'm glad that I've added that test case :)
5. I've create a GH action for this for the standard jdk versions and archs.
6. I've created a top level `bb.edn` to help with  a couple of dev tasks.

I've also noticed an issue with https://github.com/mbjarland/clj-zip-meta that when `:skip-realign false`, then I cannot call the resulting executable from inside the same test. Probably it keeps a file lock that it never released? I'll see if can find some time to look into this, but it shouldn't affect the current prod use case, whereby there is a separate process used to generate the script:

```
 Caused by java.io.IOException
   error=26, Text file busy

          ProcessImpl.java:   -2  java.lang.ProcessImpl/forkAndExec
          ProcessImpl.java:  314  java.lang.ProcessImpl/<init>
          ProcessImpl.java:  244  java.lang.ProcessImpl/start
       ProcessBuilder.java: 1110  java.lang.ProcessBuilder/start
       ProcessBuilder.java: 1073  java.lang.ProcessBuilder/start
              process.cljc:  350  babashka.process$process/invokeStatic
              process.cljc:  282  babashka.process$process/invoke
              process.cljc:  328  babashka.process$process/invokeStatic
              process.cljc:  282  babashka.process$process/invoke
              bin_test.clj:   47  deps-bin.impl.bin-test/fn/fn
```

thanks

ikappaki committed:
Add linter as bb task and enable in CI

ikappaki commented:
I've just also added a lint task and updated the test to fix linter errors and simplify the handling of process output.

ericdallo commented on bb.edn:
```diff
@@ -0,0 +1,33 @@
+             [clojure.string :as str]]
+
+  lint {:doc "Run linter."
+        :task (do (when-not (fs/exists? ".clj-kondo")
+                    ;; Prep linter. The `:test` alias has main as dep.
+                    (let [cp (with-out-str (deps/clojure ["-A:test" "-Spath"]))]
+                      (fs/create-dir ".clj-kondo")
+                      (p/shell "clj-kondo" "--lint" cp "--dependencies" "--copy-configs")))
+                  (doseq [dir ["src" "test"]]
+                    (println :linting dir "...")
+                    (p/shell "clj-kondo" "--lint" dir)
+                    (println)))}
```
We could use clojure-lsp itself to lint :) We would have clean-ns and format additional tasks
Not blocker for this PR though

ikappaki replied:
Indeed! It didn't cross my mind that you can use clojure-lsp as a command line linter, this is a really great!

I've tried it and it doesn't work on MS-Windows yet, there are a lot of false positives reported; these are fixed in the clojure-lsp patches I've provided recently pending a release.

Btw, how would you about handling definitions that are not reference anywhere, even in tests (this is on linux)?

``` shell
~/src/deps-bin $ bb lint
:linter diagnostics ...
[ 20%] Copying kondo configs       Configs copied:
- .clj-kondo/babashka/fs
- .clj-kondo/rewrite-clj/rewrite-clj
[ 99%] Analyzing project files     No configs copied.
[100%] Project analyzed            
Finding diagnostics...
src/deps_bin/deps_bin.clj:4:6: info: [clojure-lsp/unused-public-var] Unused public var 'deps-bin.deps-bin/bin'
```

It's a simple change to switch to clojure-lsp, and I've noticed you have developed a GH action https://github.com/clojure-lsp/setup-clojure-lsp too :)

bb.edn change
``` clojure
lint {:doc "Run clojure-lsp linter."
        :task (doseq [linter '[clean-ns format diagnostics]]
                (println :linter linter "...")
                (p/shell "clojure-lsp" linter "--dry" "--project-root" ".")
                (println))}
```

ericdallo replied:
We can wait for the next release, no problem
About the unused-public-var, we could use `#_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]}`

ericdallo reviewed (approved):
Looks good!
